cachito/workers/tasks/pip.py

```diff
@@ -108,3 +119,55 @@ def fetch_pip_source(request_id, package_configs=None):
 
     if pip_config_files:
         update_request_with_config_files(request_id, pip_config_files)
+
+
+def _get_custom_requirement_config_file(requirement_file_path, source_dir, raw_repo_name):
+    """
+    Get custom pip requirement file.
+
+    Generates and returns a configuration file representing a custom pip requirement file where the
+    original URL and VCS entries are replaced with entries pointing to those resources in Nexus.
+
+    :param str requirement_file_path: path to the requirement file
+    :param Path source_dir: path to the application source code
+    :param str raw_repo_name: name of the raw hosted Nexus repository containing the
+        requirements
+    :return: Cachito configuration file representation containing the custom requirement file
+    :rtype: dict
+    """
+    original_requirement_file = PipRequirementsFile(requirement_file_path)
+    cachito_requirements = []
+    differs_from_original = False
+    for requirement in original_requirement_file.requirements:
+        raw_component_name = get_raw_component_name(requirement)
+        if raw_component_name:
+            # increase max_attempts to make sure the package upload/setup is complete
+            new_url = nexus.get_raw_component_asset_url(
+                raw_repo_name, raw_component_name, max_attempts=5
+            )
+            requirement = requirement.copy(url=new_url)
+            differs_from_original = True
+
+        cachito_requirements.append(requirement)
+
+    if not differs_from_original:
+        # No vcs or url dependencies. No need for a custom requirements file
+        return
+
+    cachito_requirement_file = PipRequirementsFile.from_requirements_and_options(
+        cachito_requirements, original_requirement_file.options
+    )
+    final_contents = []
+    if cachito_requirement_file.options:
+        final_contents.append(" ".join(cachito_requirement_file.options))
+
+    final_contents.extend(
+        [str(requirement) for requirement in cachito_requirement_file.requirements]
+    )
+    req_str = "\n".join(final_contents)
+    final_path = Path("app") / Path(requirement_file_path).relative_to(source_dir)
```

Comment:
Is `source_dir` here the root of the repo? If the user specifies a subpath, wouldn't this be incorrect? I'm trying this with a repo where everything is under `subpath/`, and the config file I get has `"path": "app/requirements.txt"` when it should be `app/subpath/requirements.txt`

Reply:
source_dir here was the root of the package. I changed it to be the root of the repository so the `final_path` will be correct.